### PR TITLE
Qp modulation

### PR DESCRIPTION
**Pull request by anaghdin**

### CODE REVIEW

**Comment by anaghdin** [comment]
Description
Change the QP of each SB using deltaq to improve efficiency (Only active in Intra frames).
Updated the -adaptive-quantization: 0 = OFF , 1 = variance base using segments , 2 = Deltaq pred efficiency (default)

Authors
@anaghdin

Type of change
New Feature

Tests and performance
Expected Average PSNR-SSIM BD-rate gain in the range of -0.4% on the AOM test set, using 4 QP values: {20, 32, 43 and 55}.

No Speed impact.

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 1]
```diff
@@ -6004,7 +6004,12 @@ assert(bsize < BlockSizeS_ALL);
             blkOriginY >> MI_SIZE_LOG2);
 
 #if ADD_DELTA_QP_SUPPORT //PART 1
-        if (picture_control_set_ptr->parent_pcs_ptr->delta_q_present_flag) {
+        if (picture_control_set_ptr->parent_pcs_ptr->frm_hdr.delta_q_params.delta_q_present) {
+#if QPM
+            int32_t current_q_index = quantizer_to_qindex[cu_ptr->qp];
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbEntropyCoding.c]: Some lines may be made shorter

File: Source/Lib/Common/Codec/EbRateControlProcess.c [hunk 2]
```diff
@@ -3510,7 +3510,107 @@ static int adaptive_qindex_calc(
...
+    if (picture_control_set_ptr->parent_pcs_ptr->frm_hdr.delta_q_params.delta_q_present) {
+        const int bit_depth = sequence_control_set_ptr->static_config.encoder_bit_depth;
+        int active_best_quality = 0;
+        int active_worst_quality = quantizer_to_qindex[(uint8_t)sequence_control_set_ptr->qp];
+        int *kf_low_motion_minq;
+        int *kf_high_motion_minq;
+        ASSIGN_MINQ_TABLE(bit_depth, kf_low_motion_minq);
+        ASSIGN_MINQ_TABLE(bit_depth, kf_high_motion_minq);
+        double q_val, picture_q_val;
+        uint32_t me_sb_size = sequence_control_set_ptr->sb_sz;
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbRateControlProcess.c]: Some rather long lines here as well

File: Source/Lib/Common/Codec/EbEntropyCoding.c [hunk 3]
```diff
@@ -6004,7 +6004,12 @@ assert(bsize < BlockSizeS_ALL);
...
+            int32_t super_block_upper_left = (((blkOriginY >> 2) & (sequence_control_set_ptr->seq_header.sb_mi_size - 1)) == 0) && (((blkOriginX >> 2) & (sequence_control_set_ptr->seq_header.sb_mi_size - 1)) == 0);
+            if ((bsize != sequence_control_set_ptr->seq_header.sb_size || skipCoeff == 0) && super_block_upper_left) {
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbEntropyCoding.c]: Some lines may be made shorter

File: Source/Lib/Common/Codec/EbRateControlProcess.c [hunk 4]
```diff
@@ -3674,7 +3774,11 @@ void* rate_control_kernel(void *input_ptr)
                                 (int32_t)quantizer_to_qindex[sequence_control_set_ptr->static_config.max_qp_allowed],
                                 (int32_t)(qindex + delta_qindex));
                     }
+#if QPM
+                    picture_control_set_ptr->picture_qp = (uint8_t)CLIP3((int32_t)sequence_control_set_ptr->static_config.min_qp_allowed, (int32_t)sequence_control_set_ptr->static_config.max_qp_allowed, (frm_hdr->quantization_params.base_q_idx + 2) >> 2);
```
Comment (andrey-norkin) [inline Source/Lib/Common/Codec/EbRateControlProcess.c]: Some lines could be made shorter in this file

Comment (anaghdin) [reply Source/Lib/Common/Codec/EbRateControlProcess.c]: done

File: Source/Lib/Common/Codec/EbEntropyCoding.c (hunk 3 above)
Comment (anaghdin) [reply Source/Lib/Common/Codec/EbEntropyCoding.c]: done

File: Docs/svt-av1_encoder_user_guide.md [hunk 5]
```diff
@@ -129,6 +129,7 @@ The encoder parameters present in the `Sample.cfg` file are listed in this table
 | **IntraRefreshType** | -irefresh-type | [1 – 2] | 1 | 1: CRA (Open GOP)2: IDR (Closed GOP) |
 | **QP** | -q | [0 - 63] | 50 | Quantization parameter used when RateControl is set to 0 |
 | **RateControlMode** | -rc | [0 - 3] | 0 | 0 = CQP , 1 = ABR , 2 = ABR , 3 = CVBR |
+| **AdaptiveQuantization** | -adaptive-quantization | [0 - 2] | 0 | 0 = OFF , 1 = variance base using segments , 2 = Deltaq pred efficiency (default) |
```
Comment (hassount) [inline Docs/svt-av1_encoder_user_guide.md]: -adaptive-quantization might be too long, can we use -adapt-q

Comment (anaghdin) [reply Docs/svt-av1_encoder_user_guide.md]: It has been introduced before and I just modified it. If we all agree, I will modify it.

**Pull request merged**